core/src/main/java/org/apache/iceberg/DeleteFileIndex.java

```diff
@@ -474,68 +421,52 @@ private Collection<DeleteFile> loadDeleteFiles() {
     DeleteFileIndex build() {
       Iterable<DeleteFile> files = deleteFiles != null ? filterDeleteFiles() : loadDeleteFiles();
 
-      boolean useColumnStatsFiltering = false;
+      EqualityDeletes globalDeletes = new EqualityDeletes();
+      PartitionMap<EqualityDeletes> eqDeletesByPartition = PartitionMap.create(specsById);
+      PartitionMap<PositionDeletes> posDeletesByPartition = PartitionMap.create(specsById);
+      CharSequenceMap<PositionDeletes> posDeletesByPath = CharSequenceMap.create();
 
-      // build a map from (specId, partition) to delete file entries
-      Map<Integer, StructLikeWrapper> wrappersBySpecId = Maps.newHashMap();
-      ListMultimap<Pair<Integer, StructLikeWrapper>, IndexedDeleteFile> deleteFilesByPartition =
-          Multimaps.newListMultimap(Maps.newHashMap(), Lists::newArrayList);
       for (DeleteFile file : files) {
-        int specId = file.specId();
-        PartitionSpec spec = specsById.get(specId);
-        StructLikeWrapper wrapper =
-            wrappersBySpecId
-                .computeIfAbsent(specId, id -> StructLikeWrapper.forType(spec.partitionType()))
-                .copyFor(file.partition());
-        IndexedDeleteFile indexedFile = new IndexedDeleteFile(spec, file);
-        deleteFilesByPartition.put(Pair.of(specId, wrapper), indexedFile);
-
-        if (!useColumnStatsFiltering) {
-          useColumnStatsFiltering = indexedFile.hasLowerAndUpperBounds();
+        CharSequence path = ContentFileUtil.referencedDataFile(file);
+        if (path != null) {
+          PositionDeletes deletes = posDeletesByPath.computeIfAbsent(path, PositionDeletes::new);
+          deletes.add(file);
+        } else {
+          int specId = file.specId();
+          PartitionSpec spec = specsById.get(specId);
+          StructLike partition = file.partition();
+          if (file.content() == FileContent.POSITION_DELETES) {
+            PositionDeletes deletes = fetchPosDeletes(posDeletesByPartition, specId, partition);
```

Comment:
fetch sounds like it will load from disk, I think removing the helper method in favor of the underlying code is clearer, or renaming the method.

Reply:
I restructured this part quite a bit. Let me know if it is any better.